### DISCUSSION TITLE
Make field required dependent on other field value

### DISCUSSION THREAD

**Discussion by knoefel** [colinhacks/zod#1020] [Q&A]
I have the following schema:

```javascript
export const schema = z
  .object({
    isOfferAvailable: z.boolean(),
    offerNotAvailableReason: z.string().optional(),
  })
```

So what i want here is, when `isOfferAvailable` is `true`, then `offerNotAvailableReason` should be required and also not be empty. I've read in other discussions here that i should use `refine`. But then i have to define the refinement on the root schema object and i can't set an error message on the `offerNotAvailableReason` field directly.

In `yup` there is `when`, where i can access another value from the schema, how can i achieve this with `zod`?

Sorry if this has been asked before, i searched through the discussions and issues and couldn't find anything which solved my problem. Thx in advance!!

**Comment by scotttrinh** [comment]
One way to approach this is to use an explicit discriminated union. Not sure if the requirement is that if it's `true` that it must be `undefined` or if you just ignore that, but here it is with an explicit `undefined`.

[CodeSandbox](https://codesandbox.io/s/busy-dewdney-grk5bo?file=/src/index.ts)

```ts
const offerIsAvailable = z.object({
  isOfferAvailable: z.literal(true),
  offerNotAvailableReason: z.undefined()
});
const offerNotAvailable = z.object({
  isOfferAvailable: z.literal(false),
  offerNotAvailableReason: z.string().min(1)
});
const schema = z.discriminatedUnion("isOfferAvailable", [
  offerIsAvailable,
  offerNotAvailable
]);

test("valid", (t) => {
  t.true(schema.safeParse({ isOfferAvailable: true }).success);
  t.true(
    schema.safeParse({
      isOfferAvailable: false,
      offerNotAvailableReason: "It's around here somewhere..."
    }).success
  );
  t.end();
});

test("invalid", (t) => {
  t.false(
    schema.safeParse({
      isOfferAvailable: true,
      offerNotAvailableReason: "Oh no!"
    }).success
  );
  t.end();
});
```
